pkg/codec/x264/x264.go

```diff
@@ -0,0 +1,159 @@
+// Package x264 implements H264 encoder.
+// This package requires libx264 headers and libraries to be built.
+// Reference: https://code.videolan.org/videolan/x264/blob/master/example.c
+package x264
+
+// #cgo pkg-config: x264
+// #include "bridge.h"
+import "C"
+import (
+	"errors"
+	"fmt"
+	"image"
+	"io"
+	"sync"
+	"unsafe"
+
+	"github.com/pion/mediadevices/pkg/codec"
+	mio "github.com/pion/mediadevices/pkg/io"
+	"github.com/pion/mediadevices/pkg/io/video"
+	"github.com/pion/mediadevices/pkg/prop"
+	"github.com/pion/webrtc/v2"
+)
+
+type encoder struct {
+	engine *C.Encoder
+	buff   []byte
+	r      video.Reader
+	mu     sync.Mutex
+	closed bool
+}
+
+type cerror int
+
+func (e cerror) Error() string {
+	switch e {
+	case C.ERR_DEFAULT_PRESET:
+		return errDefaultPreset.Error()
+	case C.ERR_APPLY_PROFILE:
+		return errApplyProfile.Error()
+	case C.ERR_ALLOC_PICTURE:
+		return errAllocPicture.Error()
+	case C.ERR_OPEN_ENGINE:
+		return errOpenEngine.Error()
+	case C.ERR_ENCODE:
+		return errEncode.Error()
+	default:
+		return "unknown error"
+	}
+}
+
+func errFromC(rc C.int) error {
+	if rc == 0 {
+		return nil
+	}
+	return cerror(rc)
+}
+
+var (
+	errInitEngine    = fmt.Errorf("failed to initialize x264")
+	errDefaultPreset = fmt.Errorf("failed to set default preset")
+	errApplyProfile  = fmt.Errorf("failed to apply profile")
+	errAllocPicture  = fmt.Errorf("failed to alloc picture")
+	errOpenEngine    = fmt.Errorf("failed to open x264")
+	errEncode        = fmt.Errorf("failed to encode")
+)
+
+func init() {
+	codec.Register(webrtc.H264, codec.VideoEncoderBuilder(newEncoder))
+}
+
+func newEncoder(r video.Reader, p prop.Media) (io.ReadCloser, error) {
+	if p.KeyFrameInterval == 0 {
+		p.KeyFrameInterval = 60
+	}
+
+	var preset Preset
+	switch cp := p.CodecParams.(type) {
+	case nil:
+	case Params:
+		preset = cp.Preset
+	default:
+		return nil, errors.New("unsupported CodecParams type")
+	}
+
+	param := C.x264_param_t{
+		i_csp:        C.X264_CSP_I420,
+		i_width:      C.int(p.Width),
+		i_height:     C.int(p.Height),
+		i_keyint_max: C.int(p.KeyFrameInterval),
+	}
+	param.rc.i_bitrate = C.int(p.BitRate)
+	param.rc.i_vbv_max_bitrate = param.rc.i_bitrate
+	param.rc.i_vbv_buffer_size = param.rc.i_vbv_max_bitrate * 2
+
+	var rc C.int
+	cPreset := C.CString(fmt.Sprint(preset))
```

Comment:
maybe nice to have a comment that `cPreset` is freed in `enc_new`.

Reply:
ah, yep